src/System.Net.Utilities/src/System/Net/NetworkInformation/Ping.Windows.cs

```diff
@@ -0,0 +1,330 @@
+                }
+                else
+                {
+                    IPEndPoint ep = new IPEndPoint(address, 0);
+                    Internals.SocketAddress remoteAddr = IPEndPointExtensions.Serialize(ep);
+                    byte[] sourceAddr = new byte[28];
+
+                    SafeWaitHandle pingEventSafeWaitHandle = _pingEvent.GetSafeWaitHandle();
+                    error = (int)Interop.IpHlpApi.Icmp6SendEcho2(
+                        _handlePingV6,
+                        pingEventSafeWaitHandle,
+                        IntPtr.Zero,
+                        IntPtr.Zero,
+                        sourceAddr,
+                        remoteAddr.Buffer,
+                        _requestBuffer,
+                        (ushort)buffer.Length,
+                        ref ipOptions,
+                        _replyBuffer,
+                        MaxUdpPacket,
+                        (uint)timeout);
+                }
+            }
+            catch
+            {
+                UnregisterWaitHandle();
+                throw;
+            }
+
+            if (error == 0)
+            {
+                error = Marshal.GetLastWin32Error();
+
+                // Only skip Async IO Pending error value.
+                if (error != Interop.IpHlpApi.ERROR_IO_PENDING)
+                {
+                    // Cleanup.
+                    FreeUnmanagedStructures();
+                    UnregisterWaitHandle();
+
+                    throw new Win32Exception(error);
+                }
+            }
+        }
+
+        private void InternalDisposeCore()
+        {
+            if (_handlePingV4 != null)
+            {
+                _handlePingV4.Dispose();
+                _handlePingV4 = null;
+            }
+
+            if (_handlePingV6 != null)
+            {
+                _handlePingV6.Dispose();
+                _handlePingV6 = null;
+            }
+
+            UnregisterWaitHandle();
+
+            if (_pingEvent != null)
+            {
+                _pingEvent.Dispose();
+                _pingEvent = null;
+            }
+
+            if (_replyBuffer != null)
+            {
+                _replyBuffer.Dispose();
+                _replyBuffer = null;
+            }
+        }
+
+        static partial void InitializeSockets()
+        {
+            if (!Volatile.Read(ref s_socketInitialized))
+            {
+                lock (s_socketInitializationLock)
+                {
+                    if (!s_socketInitialized)
+                    {
+                        // Ensure that WSAStartup has been called once per process.  
+                        // The System.Net.NameResolution contract is responsible with the initialization.
+                        Dns.GetHostName();
+
+                        // Cache some settings locally.
+                        s_socketInitialized = true;
+                    }
+                }
+            }
+        }
+
+        // -----------------------------
```

Comment:
> I think now that the purpose is to clearly identify and separate the PAL interface. 

Yup.

> To clarify the code intent here we could create a different class (PingPAL) and extract an interface

That works in some cases.  It doesn't work very well here.  There is state that lives in the shared code, and then state specific to the platform-specific implementation.  If the platform-specific implementation lives in a separate class that implements an interface, then a) you need to allocate that separate class, b) that class needs to have access to the shared state in the main class (which means that state needs to have at least internal visibility rather than private), and c) the platform-specific object needs to store a reference to that main class in order to call back to it to access the shared state.  Feasible?  Sure, but it's more costly and more complex.

Alternatively you can have a platform-agnostic base class with platform-specific derived classes, but then the base still needs to expose its state as protected, you still need partial classes to ensure the right object gets allocated, you still need to allocate that inner object, etc.

With this approach, you still get the shared vs Windows vs Unix separation, you don't have any extra objects, you don't have extra interfaces or base classes, and you end up with basically exactly the same compiled code you would have written if you were targeting only the one platform.  The primary oddity is that visibility isn't usable as a means for determining what's exposed from the platform-specific to the platform-agnostic code, which I believe I've reasonably overcome via this separation in the file.

This last approach is what we've used in most of the Windows vs Unix implementations elsewhere in corefx, and it's worked out pretty well.  Unless you have strong objections, I'd like to keep it this way.   If the comment about the separation still bothers you, I can of course remove it. :smile:

Thoughts?